### PR TITLE
feat(cdc): persist the backfill state for table-on-source

### LINKED ISSUE #13204: Persist the backfill state to record backfill progress
https://github.com/risingwavelabs/risingwave/pull/12535 doesn’t persist the backfill state yet, it will backfill again during recovery.

### PR DESCRIPTION
**Pull request by StrikeW**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
Add a state table (`table_id | pk columns... | backfill_finished | row_count | cdc_offset`) to the CdcBackfill executor to persist the backfill progress during backfill. The `pk columns` would be multiple columns if the table has composite primary keys.

close #13204 

## Checklist

- [ ] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/connector/src/parser/mod.rs [hunk 1]
```diff
@@ -543,14 +543,12 @@ async fn into_chunk_stream<P: ByteStreamSourceParser>(mut parser: P, data_stream
...
             if msg.key.is_none() && msg.payload.is_none() {
-                if parser.parser_format() == ParserFormat::Debezium {
-                    tracing::debug!(offset = msg.offset, "skip parsing of heartbeat message");
-                    // empty payload means a heartbeat in cdc source
-                    // heartbeat message offset should not overwrite data messages offset
-                    split_offset_mapping
-                        .entry(msg.split_id)
-                        .or_insert(msg.offset.clone());
-                }
+                tracing::debug!(offset = msg.offset, "skip parsing of heartbeat message");
```
Comment (kwannoel) [inline src/connector/src/parser/mod.rs:546]: Seems unrelated to the PR.

File: src/connector/src/parser/mod.rs [hunk 2]
```diff
@@ -543,14 +543,12 @@ async fn into_chunk_stream<P: ByteStreamSourceParser>(mut parser: P, data_stream
...
+                // assumes an empty message as a heartbeat
+                // heartbeat message offset should not overwrite data messages offset
+                split_offset_mapping
+                    .entry(msg.split_id)
+                    .or_insert(msg.offset.clone());
```
Comment (kwannoel) [inline src/connector/src/parser/mod.rs:551]: Changes seems unrelated to the PR

File: src/frontend/src/optimizer/plan_node/stream_table_scan.rs [hunk 3]
```diff
@@ -189,6 +189,30 @@ impl StreamTableScan {
...
+    /// schema: | `table_id` | `backfill_finished` | `cdc_offset` |
+    pub fn build_cdc_backfill_state_catalog(
+        &self,
+        state: &mut BuildFragmentGraphState,
+    ) -> TableCatalog {
+        let properties = self.ctx().with_options().internal_table_subset();
+        let mut catalog_builder = TableCatalogBuilder::new(properties);
+
+        // use `table_id` as primary key in state table.
+        catalog_builder.add_column(&Field::with_name(DataType::Int64, "table_id"));
```
Comment (kwannoel) [inline src/frontend/src/optimizer/plan_node/stream_table_scan.rs:204]: Perhaps this should be associated with `TableId::RW_TYPE`. Similar to how we do for vnodes: `VirtualNode::RW_TYPE`.

File: src/connector/src/parser/mod.rs (hunk 2 above)
Comment (StrikeW) [reply src/connector/src/parser/mod.rs:551]: Actually it is related, since the schema of the cdc Source job is fixed to `FORMAT PLAIN ENCODE JSON` not `Debezium`, and we also need the heartbeat event to update the Source offset.

File: src/frontend/src/optimizer/plan_node/stream_table_scan.rs (hunk 3 above)
Comment (StrikeW) [reply src/frontend/src/optimizer/plan_node/stream_table_scan.rs:204]: Considering we may support parallel cdc backfill, perhaps we should generalize the `table_id` column to something like `split_id`.

**Comment by kwannoel** [comment]
Is there some test for cdc backfill over recovery?

**Comment by StrikeW** [comment]
> Is there some test for cdc backfill over recovery?

Right now only has e2e test. I insert new data into MV created in `cdc.share_stream.slt` after the cluster has been shutdown, then restart the cluster and check that new data has been ingested.

File: ci/scripts/e2e-source-test.sh [hunk 4]
```diff
@@ -71,10 +71,17 @@ export MYSQL_HOST=mysql MYSQL_TCP_PORT=3306 MYSQL_PWD=123456
...
 cargo make kill
 echo "cluster killed "
 
+# insert into mytest database (cdc.share_stream.slt)
+mysql --protocol=tcp -u root mytest -e "INSERT INTO products
+       VALUES (default,'RisingWave','Next generation Streaming Database'),
+              (default,'Materialize','The Streaming Database You Already Know How to Use');
+       INSERT INTO orders VALUES (default, '2022-12-01 15:08:22', 'Sam', 1000.52, 110, false);"
+
+
```
Comment (kwannoel) [inline ci/scripts/e2e-source-test.sh:84]: Can test other things like `DELETE`, `UPDATE`, and increase the workload as well.

File: src/stream/src/executor/backfill/cdc/state.rs [hunk 5]
```diff
@@ -12,71 +12,165 @@
...
+                    is_finished: finished,
+                    ..Default::default()
+                })
+            }
+            None => Ok(CdcStateItem::default()),
+        }
+    }
+
+    /// Mark the backfill has done and save the last cdc offset
+    pub async fn mutate_state(&mut self, state_item: CdcStateItem) -> StreamExecutorResult<()> {
```
Comment (kwannoel) [inline src/stream/src/executor/backfill/cdc/state.rs:132]:
```suggestion
    pub async fn finish_state(&mut self, state_item: CdcStateItem) -> StreamExecutorResult<()> {
```

File: src/stream/src/executor/backfill/cdc/state.rs [hunk 6]
```diff
@@ -12,71 +12,165 @@
...
 use risingwave_connector::source::external::{CdcOffset, DebeziumOffset, DebeziumSourceOffset};
 use risingwave_connector::source::{SplitId, SplitImpl, SplitMetaData};
 use risingwave_storage::StateStore;
 use serde_json::Value;
 
+use crate::common::table::state_table::StateTable;
 use crate::executor::{SourceStateTableHandler, StreamExecutorResult};
 
+#[allow(dead_code)]
 pub enum CdcBackfillStateImpl<S: StateStore> {
```
Comment (kwannoel) [inline src/stream/src/executor/backfill/cdc/state.rs:30]: Can we add some docs here about the different states for cdc backfill?

File: src/stream/src/executor/backfill/cdc/state.rs [hunk 7]
```diff
@@ -12,71 +12,165 @@
...
+        {
+            Some(row) => {
+                let finished = match row.datum_at(1) {
+                    Some(ScalarRefImpl::Bool(val)) => val,
+                    _ => return Err(anyhow!("invalid backfill state: backfill_finished").into()),
+                };
+
+                Ok(CdcStateItem {
+                    is_finished: finished,
+                    ..Default::default()
```
Comment (kwannoel) [inline src/stream/src/executor/backfill/cdc/state.rs:124]: Hmm should we only recover `is_finished`? What about last cdc offset, row count etc

File: src/stream/src/executor/backfill/cdc/state.rs [hunk 8]
```diff
@@ -12,71 +12,165 @@
...
+                })
+            }
+            None => Ok(CdcStateItem::default()),
+        }
+    }
+
```
Comment (kwannoel) [inline src/stream/src/executor/backfill/cdc/state.rs:130]: Should there be some other function to regularly update the backfill state?
Seems only have functions which restore, and finish the backfill state.
While backfilling, I suppose on barrier we should checkpoint the state?

File: src/stream/src/executor/backfill/cdc/state.rs [hunk 9]
```diff
@@ -95,26 +189,27 @@ impl<S: StateStore> SingleTableState<S> {
...
 
     /// When snapshot read stream ends, we should persist two states:
     /// 1) a backfill finish flag to denote the backfill has done
     /// 2) a consumed binlog offset to denote the last binlog offset
     /// which will be committed to the state store upon next barrier.
-    pub async fn mutate_state(
-        &mut self,
-        last_binlog_offset: Option<CdcOffset>,
-    ) -> StreamExecutorResult<()> {
+    pub async fn mutate_state(&mut self, state_item: CdcStateItem) -> StreamExecutorResult<()> {
```
Comment (kwannoel) [inline src/stream/src/executor/backfill/cdc/state.rs:212]:
```suggestion
    pub async fn finish_state(&mut self, state_item: CdcStateItem) -> StreamExecutorResult<()> {
```

File: ci/scripts/e2e-source-test.sh (hunk 4 above)
Comment (StrikeW) [reply ci/scripts/e2e-source-test.sh:85]: Update and I found a bug in mview exec https://github.com/risingwavelabs/risingwave/issues/13346

**Comment by StrikeW** [comment]
Existing comments addressed. PTAL @hzxa21 @fuyufjh @BugenZhao

File: src/frontend/src/optimizer/plan_node/stream_table_scan.rs [hunk 10]
```diff
@@ -192,6 +192,37 @@ impl StreamTableScan {
...
+        let properties = self.ctx().with_options().internal_table_subset();
+        let mut catalog_builder = TableCatalogBuilder::new(properties);
+
+        // use `table_id` as primary key in state table.
+        catalog_builder.add_column(&Field::with_name(DataType::Varchar, "split_id"));
+        catalog_builder.add_order_column(0, OrderType::ascending());
+
+        catalog_builder.add_column(&Field::with_name(
+            DataType::Boolean,
+            format!("{}_backfill_finished", self.table_name()),
```
Comment (hzxa21) [inline src/frontend/src/optimizer/plan_node/stream_table_scan.rs:212]: I suggest we don't put the external `table_name` here as the prefix of the column name so that all table's backfill state share the same schema. It is easy to construt debugging SQL query in this case.

**Review by fuyufjh — COMMENTED** [review]
Mostly LGTM.

This PR only records a boolean flag - completed or not in the persisted state, intead of a real progress (marked by PK perhaps). I think we had better complete this before releasing to users, because we can't say the progress is persisted but either 0% or 100%.

File: src/stream/src/executor/backfill/cdc/cdc_backfill.rs [hunk 11]
```diff
@@ -334,8 +343,16 @@ impl<S: StateStore> CdcBackfillExecutor<S> {
                                         last_binlog_offset = consumed_binlog_offset.clone();
                                     }
 
-                                    // seal current epoch even though there is no data
+                                    // update and persist backfill state
+                                    state_impl
+                                        .mutate_state(CdcStateItem::new(
+                                            false,
+                                            last_binlog_offset.clone(),
+                                            total_snapshot_processed_rows as _,
```
Comment (hzxa21) [inline src/stream/src/executor/backfill/cdc/cdc_backfill.rs:351]: IIUC, `last_binlog_offset` is not used after restoring the state and we don't store the pk position for mysql snapshot read. Are we planning to support resume backfill on the exact pk after recovery? If yes, will it be incompatible in the future?

File: src/stream/src/executor/backfill/cdc/state.rs [hunk 12]
```diff
@@ -12,71 +12,171 @@
...
+/// 2. `CREATE TABLE xx FROM source TABLE 'mydb.t1'`
+/// In this case, we can have multiple Table jobs sharing a single cdc Source job.
+/// The cdc backfill executor will be an instance of the `StreamScan` operator and has its own state table
+/// schema: `table_id | backfill_finished | row_count | cdc_offset`
 pub enum CdcBackfillStateImpl<S: StateStore> {
-    Undefined,
-    SingleTable(SingleTableState<S>),
+    SingleTable(SingleBackfillState<S>),
+    MultiTable(MultiBackfillState<S>),
+}
```
[the remark this answers is not in the document]
Comment (StrikeW) [reply src/stream/src/executor/backfill/cdc/state.rs:42]: I also want to deprecate it. But we need to change the original `Table` job plan from `source -> mview` to `source -> stream_scan -> mview`, so that it can has its own state, technically it is feasible.

**Review by StrikeW — COMMENTED** [review]
I updated the pr to support recoverable backfill, that is to record the backfill progress and can continue the progress upon cluster recovery.

File: src/frontend/src/optimizer/plan_node/stream_table_scan.rs [hunk 13]
```diff
@@ -192,6 +192,37 @@ impl StreamTableScan {
...
+    /// schema: | `table_id` | `backfill_finished` | `row_count` | `cdc_offset` |
+    pub fn build_cdc_backfill_state_catalog(
+        &self,
+        state: &mut BuildFragmentGraphState,
+    ) -> TableCatalog {
+        let properties = self.ctx().with_options().internal_table_subset();
+        let mut catalog_builder = TableCatalogBuilder::new(properties);
+
+        // use `table_id` as primary key in state table.
+        catalog_builder.add_column(&Field::with_name(DataType::Varchar, "split_id"));
```
[the remark this answers is not in the document]
Comment (StrikeW) [reply src/frontend/src/optimizer/plan_node/stream_table_scan.rs]: Mistake here. Fixed

File: src/frontend/src/optimizer/plan_node/stream_table_scan.rs (hunk 10 above)
Comment (StrikeW) [reply src/frontend/src/optimizer/plan_node/stream_table_scan.rs]: agree, fixed.

File: src/stream/src/executor/backfill/cdc/cdc_backfill.rs (hunk 11 above)
Comment (StrikeW) [reply src/stream/src/executor/backfill/cdc/cdc_backfill.rs]: I updated the PR to support recoverable backfill.

File: src/stream/src/executor/backfill/cdc/state.rs [hunk 14]
```diff
@@ -95,37 +220,46 @@ impl<S: StateStore> SingleTableState<S> {
...
+        // 1) a finish flag to denote the backfill has done
+        // 2) a consumed binlog offset to denote the last binlog offset
+        // which will be committed to the state store upon next barrier.
         let mut key = self.split_id.to_string();
         key.push_str(BACKFILL_STATE_KEY_SUFFIX);
         // write backfill finished flag
         self.source_state_handler
-            .set(key.into(), JsonbVal::from(Value::Bool(true)))
+            .set(
+                key.into(),
```
[the remark this answers is not in the document]
Comment (StrikeW) [reply src/stream/src/executor/backfill/cdc/state.rs:254]: Since for the backfill for single table scenario, we didn't introduce a new operator to the query plan. Just wraps the source executor in the cdc backfill executor and reuse the state table of `Source`. We access source state table only via point query with the ke column, so it won't break.

[the remark this answers is not in the document]
Comment (hzxa21) [reply src/stream/src/executor/backfill/cdc/state.rs:254]: True. If only get is used in the streaming job, there won't be unexpected result. But when querying source state via SQL, we can see something like this:
```
partition_id  |                                                                                                                                                                      offset_info                                                                                                                                                                       
---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 1002          | {"split_info": {"mysql_split": {"inner": {"snapshot_done": false, "split_id": 1002, "start_offset": "{\"sourcePartition\":{\"server\":\"RW_CDC_1002\"},\"sourceOffset\":{\"transaction_id\":null,\"ts_sec\":1699866101,\"file\":\"binlog.000012\",\"pos\":157,\"server_id\":1},\"isHeartbeat\":true}"}}, "pg_split": null}, "split_type": "mysql-cdc"}
 1002_backfill | true
(2 rows)
```

This looks very strange to me. How about storing the backfill finished flag inside the json as a new field under "mysql_split"? This issue is introduced in #12535 and not related to the changes in this PR. I am okay with handling it separately.

[the remark this answers is not in the document]
Comment (StrikeW) [reply src/stream/src/executor/backfill/cdc/state.rs:254]: Yeah, it is a bit dirty. As mentioned in https://github.com/risingwavelabs/risingwave/pull/13276#discussion_r1390589039, I plan to refactor the plan of  single table cdc backfill to make the cdc backfill become a separate operator instead of just wraps the source inside it.

File: src/stream/src/executor/backfill/cdc/state.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (hzxa21) [reply src/stream/src/executor/backfill/cdc/state.rs:42]: +1. Currently we have 3 ways to use mysql-cdc:
- `create table xxx with (....)`
- `set cdc_backfill='true'` + `create table xxx with (...)`
- `set cdc_backfill=true` + `create source with (...)` + `create table from ...`

It is confused and not easy to explain to user which one to choose. Ideally we should only have 2 ways:
1. `create table xxx with (....)`
2. `create source with (...)` + `create table from ...`

1 and 2 can share the same implementation and 1 can be a syntax suger for 2 if only user only needs to cdc one table from the db.

[the remark this answers is not in the document]
Comment (hzxa21) [reply src/stream/src/executor/backfill/cdc/state.rs:42]: The question is whether we need to ensure compatibility to make existing jobs created via set `cdc_backfill='true'` + `create table xxx with (...)` to be compatible. IMO, we can keep it simple and avoid compatibility codes here since cdc_backfill is announced as experimental feature only.

**Comment by kwannoel** [comment]
> I updated the pr to support recoverable backfill, that is to record the backfill progress and can continue the progress upon cluster recovery.

Could you update the PR description and add a test for this as well (separate PR is fine as well, as long as there's an issue to track it).

Previously the test only tests for case where cdc backfill is done.

**Comment by kwannoel** [comment]
Also do we need both pk and cdc offset?
If we have pk already, seems we don't need cdc offset?

**Comment by StrikeW** [comment]
Thanks for the reminder, I will add more test for the recoverable backfill part, tracked by https://github.com/risingwavelabs/risingwave/issues/13400

> Also do we need both pk and cdc offset? If we have pk already, seems we don't need cdc offset?

The persisted cdc_offset acts as a low watermark for upstream events to ignore events with offset that less than the cdc_offset. And also may help us debuging the backfill process.

File: src/stream/src/from_proto/stream_scan.rs [hunk 15]
```diff
@@ -130,11 +130,11 @@ impl ExecutorBuilder for ChainExecutorBuilder {
                     (0..table_desc.columns.len()).collect_vec(),
                 );
 
-                let source_state_handler = SourceStateTableHandler::from_table_catalog(
-                    node.get_state_table().as_ref().unwrap(),
-                    state_store.clone(),
-                )
-                .await;
+                let vnodes = params.vnode_bitmap.map(Arc::new);
```
[the remark this answers is not in the document]
Comment (fuyufjh) [reply src/stream/src/from_proto/stream_scan.rs:104]: Recommend to add an assertion here

File: src/meta/service/src/ddl_service.rs [hunk 16]
```diff
@@ -457,25 +458,22 @@ impl DdlService for DdlServiceImpl {
...
-                ) {
-                    fill_cdc_mysql_server_id(&mut fragment_graph);
-                }
-            }
-        }
 
-        let mut stream_job = StreamingJob::Table(source, mview, job_type);
+        // Generate source id.
+        let source_id = if source.is_some() {
+            self.gen_unique_id::<{ IdCategory::Table }>().await? // TODO: Use source category
```
[the remark this answers is not in the document]
Comment (fuyufjh) [reply src/meta/service/src/ddl_service.rs:465]: I see.

**Review by fuyufjh — APPROVED** [review]
LGTM
